src/Framework/BuildException/BuildExceptionSerializationHelper.cs

```diff
@@ -0,0 +1,273 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using Microsoft.Build.Shared;
+
+namespace Microsoft.Build.BackEnd
+{
+    internal static class BuildExceptionSerializationHelper
+    {
+        private class BuildExceptionConstructionCallbacks
+        {
+            public BuildExceptionConstructionCallbacks(
+                Func<string, Exception?, Exception> factory,
+                Action<Exception, BuildExceptionRemoteState> instnaceInitializer,
+                Func<Exception, IDictionary<string, string?>?> remoteStateExtractor)
+            {
+                Factory = factory;
+                InstnaceInitializer = instnaceInitializer;
+                RemoteStateExtractor = remoteStateExtractor;
+            }
+
+            internal Func<string, Exception?, Exception> Factory { get; }
+            internal Action<Exception, BuildExceptionRemoteState> InstnaceInitializer { get; }
+            internal Func<Exception, IDictionary<string, string?>?> RemoteStateExtractor { get; }
+        }
+
+        private static readonly Dictionary<string, BuildExceptionConstructionCallbacks> s_exceptionFactories = FetchExceptionsConstructors();
```

Comment:
This should actually be populated on-demand, rather then prefetched - this way we can prevent problematic situations with lately loaded assemblies.
In order to load on-demand on the receiving side, we'll need to send type FQDN.